### PR TITLE
Automatic sample set downloader

### PR DESCRIPTION
**Pull request by louiseichhorst**

### CODE REVIEW

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
+//=============================================================================================================
+/**
+* @file     Downloader.cpp
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: change do downloader.cpp

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+*
+*/
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "Downloader.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: this line is what causes travis to fail. change to #include "downloader.h"

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+#include "Extract.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: see above

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+#include "ui_downloader.h"
+
+//*************************************************************************************************************
+//=============================================================================================================
+// QT INCLUDES
+//=============================================================================================================
+
+#include <QFileDialog>
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: add linebreak

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+#ifdef _WIN32
+
+void Downloader::on_extractButton_clicked()
+{
+    QString zipPath = "\"" + QString(QDir::toNativeSeparators(ui->filePath->text())) + "\"";
+    readyToExtract(boolDownloadStatus, zipPath, m_qCurrentPath);
+    ui->extractButton->setEnabled(false);
+
+}
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: Add spacer here

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    ui->extractButton->setEnabled(true);
+    ui->downloadButton->setVisible(false);
+    ui->downloadButton->setEnabled(false);
+    ui->progressBar->setVisible(false);
+    ui->filePath->setVisible(true);
+    ui->toolButton->setVisible(true);
+    ui->toolButton->setEnabled(true);
+}
+
+void Downloader::on_toolButton_clicked()
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: please do not use underscores in function names. change to onToolButtonClicked. 

http://wiki.mne-cpp.org/index.php/Coding_Conventions_%26_Style

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+#endif
+
+void Downloader::done()
+{
+    ui->label->setText("done");
+}
+
+//placeholder downloadFinished OSX
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: Change this to TODO: Add OSX support. QtCreator can conveniently scan for TODO strings in our project files. This way we can always generate a To Do list from our code when using the TODO syntax

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+void Downloader::dataReady()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    return;
+}
+
+void Downloader::downloadProgress(qint64 recieved,qint64 total)
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: please use one space between qint64 recieved and qint64 total: qint64 recieved, qint64 total

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    m_qFile.open(QIODevice::Append);
+
+    connect(m_qReply.data(), SIGNAL(downloadProgress(qint64,qint64)), this, SLOT(downloadProgress(qint64,qint64)));
+    QEventLoop loop;
+    connect(m_qReply.data(), SIGNAL(finished()), &loop, SLOT(quit()));
+    loop.exec();
+    return;
+}
+
+void Downloader::on_exitButton_clicked()
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: see above comment regarding underscores

File: applications/mne_sample_set_downloader/downloader.cpp [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+{
+    this->close();
+    m_qFile.remove();
+    m_qFile.close();
+    return;
+}
+
+
+Downloader::Downloader(QWidget *parent) :
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.cpp]: Please keep the same order of the functions in the cpp file as in the header file

File: applications/mne_sample_set_downloader/downloader.h [hunk 11]
```diff
@@ -0,0 +1,141 @@
+//=============================================================================================================
+/**
+* @file     mne_sample_set_downloader.h
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: this is not the name of the file

File: applications/mne_sample_set_downloader/downloader.h [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+#ifndef DOWNLOADER_H
+#define DOWNLOADER_H
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "Extract.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: see above regarding correct file inclusion: #include "extract.h"

File: applications/mne_sample_set_downloader/downloader.h [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+#include <QNetworkReply>
+#include <QUrl>
+#include <QString>
+#include <QtCore>
+#include <QFile>
+#include <QSharedPointer>
+#include <QDebug>
+#include <QIODevice>
+#include <QFileDialog>
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: Add linebreak

File: applications/mne_sample_set_downloader/downloader.h [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+//=============================================================================================================
+// DEFINE NAMESPACE DOWNLOADER
+//=============================================================================================================
+
+using namespace std;
+
+namespace Ui {
+class Downloader;
+}
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: Add linebreak

File: applications/mne_sample_set_downloader/downloader.h [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+#include <QFileDialog>
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE NAMESPACE DOWNLOADER
+//=============================================================================================================
+
+using namespace std;
+
+namespace Ui {
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: Nice work on using forward declarations

File: applications/mne_sample_set_downloader/downloader.h [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+/**
+* This class builds a Downloader Object an GUI. It downloads the sample data set and stores it in a temporary file
+* for the Extract class to extract once it recieves the download_finished signal.
+* The temporary file is deleted on leaving the GUI. The class also handles error and finalisation signals
+* from the extract class and adjusts the GUI accordingly (i.e. displays "Done" on finalization and ask if/where 7zip is
+* installed if the Extract class can't find it).
+*
+* @brief Downloads the sample data set and manages the GUI
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: This is the description of the overall file. Should say: Downloader class declaration.

File: applications/mne_sample_set_downloader/downloader.h [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+*/
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: you can add the brief description from above in doxygen sytanx here

File: applications/mne_sample_set_downloader/downloader.h [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    //=========================================================================================================
+    /**
+    * Constructs a Downloader object.
+    */
+    explicit Downloader(QWidget *parent = 0);
+    bool boolDownloadStatus = false;
+    ~Downloader();
+
+public slots:
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: please do not use the slot synatx. it is deprecated. just use a simple public, protected, private keyword.

File: applications/mne_sample_set_downloader/downloader.h [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+public:
+    typedef QSharedPointer<Downloader> SPtr;            /**< Shared pointer type for Downloader. */
+    typedef QSharedPointer<const Downloader> ConstSPtr; /**< Const shared pointer type for Downloader. */
+
+    //=========================================================================================================
+    /**
+    * Constructs a Downloader object.
+    */
+    explicit Downloader(QWidget *parent = 0);
+    bool boolDownloadStatus = false;
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: variable declaration go beneath function declaration. also do not initialize variables in the header but in the constructor. also change name to m_bDownloadStatus

File: applications/mne_sample_set_downloader/downloader.h [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+    ~Downloader();
+
+public slots:
+    void done();
+
+private slots:
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: see above

File: applications/mne_sample_set_downloader/downloader.h [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    void downloadFinished();
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/downloader.h]: add doxgen documentation for each function and variable

File: applications/mne_sample_set_downloader/extract.cpp [hunk 22]
```diff
@@ -0,0 +1,118 @@
+//=============================================================================================================
+/**
+* @file     Extract.cpp
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: fix naming

File: applications/mne_sample_set_downloader/extract.cpp [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+*
+*/
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "Extract.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: fix include naming

File: applications/mne_sample_set_downloader/extract.cpp [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+{
+
+}
+
+#ifdef _WIN32
+
+
+void Extract::extractTar()
+{
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: remove empty line

File: applications/mne_sample_set_downloader/extract.cpp [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+    QProcess *m_qp7zipTar = new QProcess(this);
+    m_qArguments.removeLast();
+    m_qArguments << (m_qCurrentPath + "\\MNE-Sample-Data.tar");
+    connect(m_qp7zipTar, SIGNAL(finished(int,QProcess::ExitStatus)), m_qp7zipTar, SLOT(deleteLater()));
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: Do not use SIGNAL SLOT syntax. See here for correct usage: http://doc.qt.io/qt-5/signalsandslots.html

File: applications/mne_sample_set_downloader/extract.cpp [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+    m_qp7zipTar->start(m_q7zipPath, m_qArguments);
+    m_qp7zipTar->waitForFinished(-1);
+
+    QFile tarFile("MNE-Sample-Data.tar");
+    tarFile.remove();
+    tarFile.close();
+
+    emit extractionDone();
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: remove empty line

File: applications/mne_sample_set_downloader/extract.cpp [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void Extract::extractGz(QString archivePath)
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: use const QString& archivePath instead

File: applications/mne_sample_set_downloader/extract.cpp [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+{
+    m_qArguments.clear();
+    QFile oldData("MNE-Sample-Data.tar");
+    oldData.remove();
+    oldData.close();
+    m_qCurrentPath = archivePath;
+    m_qArguments << "x" << "-aoa" << (m_qCurrentPath + "\\sample.tar.gz");
+    QProcess *m_qp7zipGz = new QProcess(this);
+    connect(m_qp7zipGz, SIGNAL(finished(int,QProcess::ExitStatus)), m_qp7zipGz, SLOT(deleteLater()));
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: see above on correct signal slot syntax

File: applications/mne_sample_set_downloader/extract.cpp [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+    connect(m_qp7zipGz, SIGNAL(finished(int,QProcess::ExitStatus)), this, SLOT(extractTar()));
+    connect(m_qp7zipGz, SIGNAL(errorOccurred(QProcess::ProcessError)), this, SIGNAL(zipperError()));
+    m_qp7zipGz->start(m_q7zipPath, m_qArguments);
+    qDebug() << m_qp7zipGz->error();
+
+}
+
+void Extract::beginExtraction(QString zip, QString current)
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: use cosnt reference instead const QString& zip, const QString& current

File: applications/mne_sample_set_downloader/extract.cpp [hunk 30]
```diff
@@ -0,0 +1,118 @@
...
+{
+    if (zip == NULL)
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.cpp]: use {} brackets

File: applications/mne_sample_set_downloader/extract.h [hunk 31]
```diff
@@ -0,0 +1,140 @@
+//=============================================================================================================
+/**
+* @file     Extract.h
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: fix naming

File: applications/mne_sample_set_downloader/extract.h [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+//*************************************************************************************************************
+//=============================================================================================================
+// Extract FORWARD DECLARATIONS
+//=============================================================================================================
+
+
+//=============================================================================================================
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: remove empty line

File: applications/mne_sample_set_downloader/extract.h [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+*
+* @brief Extracts the sample data set.
+*/
+
+class Extract : public QMainWindow
+{
+    Q_OBJECT
+
+public:
+    typedef QSharedPointer<Extract> SPtr;            /**< Shared pointer type for mne_sample_set_downloader. */
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: docu should say /**< Shared pointer type for Extract. */

File: applications/mne_sample_set_downloader/extract.h [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+    typedef QSharedPointer<const Extract> ConstSPtr; /**< Const shared pointer type for mne_sample_set_downloader. */
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: see above

File: applications/mne_sample_set_downloader/extract.h [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+
+    //=========================================================================================================
+    /**
+    * Constructs a Extract object.
+    */
+    explicit Extract(QWidget *parent = 0);
+    ~Extract();
+
+private:
+    Ui::Extract *ui;
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: use doxygen comments for documentation

File: applications/mne_sample_set_downloader/extract.h [hunk 36]
```diff
@@ -0,0 +1,140 @@
...
+    QStringList m_qArguments;
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: Mind indentation

File: applications/mne_sample_set_downloader/extract.h [hunk 37]
```diff
@@ -0,0 +1,140 @@
...
+public:
+    void beginExtraction(QString, QString);
+
+signals:
+    void zipperError();
+
+private:
+    void extractGz(QString);
+
+private slots:
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/extract.h]: get rid of signal slots syntax

File: applications/mne_sample_set_downloader/main.cpp [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief    Main.cpp starts the program.
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/main.cpp]: change to main.cpp

File: applications/mne_sample_set_downloader/main.cpp [hunk 39]
```diff
@@ -0,0 +1,77 @@
...
+*
+*/
+
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "Downloader.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/main.cpp]: fix naming

File: applications/mne_sample_set_downloader/main.cpp [hunk 40]
```diff
@@ -0,0 +1,77 @@
...
+#include "Extract.h"
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/main.cpp]: see above

File: applications/mne_sample_set_downloader/main.cpp [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+* The function main marks the entry point of the program.
+* By default, main has the storage class extern.
+*
+* @param [in] argc (argument count) is an integer that indicates how many arguments were entered on the command line when the program was started.
+* @param [in] argv (argument vector) is an array of pointers to arrays of character objects. The array objects are null-terminated strings, representing the arguments that were entered on the command line when the program was started.
+* @return the value that was set to exit() (which is 0 if exit() is called via quit()).
+*/
+
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/main.cpp]: remove the folloing two lines

File: applications/mne_sample_set_downloader/mne_sample_set_downloader.pro [hunk 42]
```diff
@@ -0,0 +1,110 @@
...
+#
+# @file     mne_sample_set_downloader.pro
+# @author   Louis Eichhorst <louis.eichhorst@tu-ilmenau.de>
+#           Matti Hamalainen <msh@nmr.mgh.harvard.edu>
+# @version  1.0
+# @date     August, 2016
+#
+# @section  LICENSE
+#
+# Copyright (C) 2016, Christoph Dinh, Martin Luessi and Matti Hamalainen. All rights reserved.
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/mne_sample_set_downloader.pro]: fix author list

File: applications/mne_sample_set_downloader/mne_sample_set_downloader.pro [hunk 43]
```diff
@@ -0,0 +1,110 @@
...
+#
+#--------------------------------------------------------------------------------------------------------------
+
+include(../../mne-cpp.pri)
+
+VERSION = $${MNE_CPP_VERSION}
+
+CONFIG   += qt
+
+QT       += core gui widgets
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/mne_sample_set_downloader.pro]: QT       += core gui widgets network

File: applications/mne_sample_set_downloader/mne_sample_set_downloader.pro [hunk 44]
```diff
@@ -0,0 +1,110 @@
...
+
+QT       += network
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/mne_sample_set_downloader.pro]: remove

File: applications/mne_sample_set_downloader/mne_sample_set_downloader.pro [hunk 45]
```diff
@@ -0,0 +1,110 @@
...
+SOURCES += main.cpp\
+    downloader.cpp \
+    extract.cpp
+
+HEADERS  += \
+    downloader.h \
+    extract.h
+
+FORMS    += \
+    Downloader.ui
```
Comment (LorenzE) [inline applications/mne_sample_set_downloader/mne_sample_set_downloader.pro]: change to downloader.ui

File: applications/mne_sample_set_downloader/downloader.cpp (hunk 6 above)
Comment (louiseichhorst) [inline applications/mne_sample_set_downloader/downloader.cpp]: that's the standardname for a slot listening for the buttonClicked signal. If changed it won't be connected to the button on the GUI

File: applications/mne_sample_set_downloader/downloader.h (hunk 18 above)
Comment (louiseichhorst) [inline applications/mne_sample_set_downloader/downloader.h]: If left out the slots for the buttons won't be recognized as such. This causes the buttons on the GUI to lose their function

**Comment by LorenzE** [comment]
Nice work Louis! Keep it up!

**Pull request merged**